File: 프로그래머스/2/42626. 더 맵게/더 맵게.cpp
#include <bits/stdc++.h>

using namespace std;

priority_queue<int, vector<int>, greater<int>> pq;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    
    for(int i = 0; i < scoville.size(); i++){
        pq.push(scoville[i]);
    }    
    
    while(pq.top() < K){
        if(pq.size() == 1 || pq.empty()){
            answer = -1;
            break;
        }
        int a = pq.top(); pq.pop();
        int b = pq.top(); pq.pop();
        
        int new_prod = a + (b * 2);
        
        pq.push(new_prod); answer++;
    }
    
    return answer;
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.java
import java.util.*;

class Solution {
    static PriorityQueue<Integer> pq = new PriorityQueue<>();
    
    public int solution(int[] scoville, int K) {
        int answer = 0;
        int size = scoville.length;
        
        for(int i = 0; i < size; i++){
            pq.add(scoville[i]);
        }
        
        while(!pq.isEmpty()){
            int a = pq.poll();
            if(a >= K){
                return answer;
            }
            if(pq.isEmpty()){
                break;
            }
            int b = pq.poll();
            
            int mixed = a + (b * 2);
            
            answer++;
            
            pq.add(mixed);
        }
        
        return -1;
    }
}